fleetctl/fleetctl.go

```diff
@@ -726,24 +809,65 @@ func lazyCreateUnits(args []string) error {
 	return nil
 }
 
-func warnOnDifferentLocalUnit(loc string, su *schema.Unit) {
-	suf := schema.MapSchemaUnitOptionsToUnitFile(su.Options)
-	if _, err := os.Stat(loc); !os.IsNotExist(err) {
-		luf, err := getUnitFromFile(loc)
-		if err == nil && luf.Hash() != suf.Hash() {
-			stderr("WARNING: Unit %s in registry differs from local unit file %s", su.Name, loc)
-			return
+// matchLocalFileAndUnit compares a file with a Unit
+// Returns true if the contents of the file matches the unit one, false
+// otherwise; and any error encountered.
+func matchLocalFileAndUnit(file string, su *schema.Unit) (bool, error) {
+	result := false
+	a := schema.MapSchemaUnitOptionsToUnitFile(su.Options)
+
+	_, err := os.Stat(file)
+	if err == nil {
+		b, err := getUnitFromFile(file)
+		if err == nil {
+			result = unit.MatchUnitFiles(a, b)
 		}
 	}
-	if uni := unit.NewUnitNameInfo(path.Base(loc)); uni != nil && uni.IsInstance() {
-		file := path.Join(path.Dir(loc), uni.Template)
-		if _, err := os.Stat(file); !os.IsNotExist(err) {
-			tmpl, err := getUnitFromFile(file)
-			if err == nil && tmpl.Hash() != suf.Hash() {
-				stderr("WARNING: Unit %s in registry differs from local template unit file %s", su.Name, uni.Template)
-			}
+
+	return result, err
+}
+
+// isLocalUnitDifferent compares a Unit on the file system with a one
+// provided from the Registry.
+// isLocalUnitDifferent first tries to load the passed Unit from the
+// local file system and compares it with the Unit that is in the
+// Registry. If it fails to load that Unit from the filesystem and
+// fatal was not set, it will check again if that file name is an
+// instance of a template, if so it will load the template Unit and
+// compare it with the provided Unit.
+// It takes four arguments; a path to the local Unit on the file system,
+// the Unit in the registry, a boolean to warn in case the Units differ;
+// and a last boolean to fail in case fatal errors happen.
+// Returns true if the local Unit on file system is different from the
+// one provided, false otherwise; and any error encountered.
+func isLocalUnitDifferent(file string, su *schema.Unit, warnIfDifferent bool, fatal bool) (bool, error) {
+	result, err := matchLocalFileAndUnit(file, su)
+	if err == nil {
+		if result == false && warnIfDifferent {
+			stderr("WARNING: Unit %s in registry differs from local unit file %s", su.Name, file)
+		}
+		return !result, nil
+	} else if fatal {
+		return false, err
+	}
+
+	info := unit.NewUnitNameInfo(path.Base(file))
+	if info == nil {
+		return false, fmt.Errorf("error extracting information from unit name %s", file)
+	} else if !info.IsInstance() {
+		return false, fmt.Errorf("error Unit %s does not seem to be a template unit", file)
```

Comment:
Why ? 

> there is no reason to assume at this point that the Unit is supposed to be a template unit...
Sorry I don't get you at all! that's the same code before, and errors are ignored for `!Replace` cases for backward compatibility... but for replace we want to compare template units otherwise we can't replace them nor their instances...

Sorry but I don't get your comments, could you please be more specific ? Thank you!